### PR TITLE
#157721401 setup postgresql database

### PR DESCRIPTION
**Pull request by cwizard2011**
#### What does this PR do?
Setup the app to communicate with PostgreSQL

#### Description of Task to be completed?
The app should send data and retrieve data from the database

#### How should this be manually tested?
- clone this app using ```git clone https://github.com/cwizard2011/maintenace-tracker-app.git```
- navigate to the app directory using ```cd maintenace-tracker-app``` in your terminal
- run ```npm install```
- setup your database url following the guide in ```.env.example```
- In your database run query ```CREATE TYPE role AS ENUM ('users', 'admin')``` and ```CREATE TYPE status AS ENUM ('new', 'approved', 'pending', 'rejected', 'resolved') ``` to create your ENUMERATION data type
- run ```node server/models/database.js``` to create your user table and request table in the database
- check your database to verify the created table
#### Any background context you want to provide?
The app is setup using 
-node version 9.11.1,
- pg version 6.1.0
- elephantsql

#### What are the relevant pivotal tracker stories?
#157721401

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,7 +30,8 @@
     "body-parser": "^1.18.3",
     "dotenv": "^5.0.1",
     "eslint": "^4.19.1",
-    "express": "^4.16.3"
+    "express": "^4.16.3",
```
Comment (paradisekelechi) [inline package.json:34]: Do you know what the caret does?

File: server/models/database.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const pg = require('pg');
```
Comment (paradisekelechi) [inline server/models/database.js]: Why the use of `require`? Aren't you using es6?

File: server/models/database.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+const client = new pg.Client(connectionString);
+client.connect();
+const users = client.query('CREATE TABLE userlist(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), username VARCHAR(40) NOT NULL, password VARCHAR(40) NOT NULL, firstName VARCHAR(40) NOT NULL, lastName VARCHAR(40) NOT NULL, user_role role default \'users\', email VARCHAR(40) NOT NULL, UNIQUE(email, username) )');
+const request = client.query('CREATE TABLE requests(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), user_id UUID, title VARCHAR(40), currentStatus status default \'new\', details VARCHAR(255), UNIQUE(title), FOREIGN KEY(user_id) REFERENCES userlist(id) )');
+let count = 1;
+
+function endHandler() {
+  count -= 1; // decrement count by 1
+  if (count === 0) {
+    // two queries have ended, lets close the connection.
```
Comment (paradisekelechi) [inline server/models/database.js]: This document style is not in line with airbnb style

Comment (paradisekelechi) [inline server/models/database.js]: This document style is not in line with airbnb style

File: server/models/database.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+const connectionString = process.env.DATABASE_URL;
+
+const client = new pg.Client(connectionString);
+client.connect();
+const users = client.query('CREATE TABLE userlist(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), username VARCHAR(40) NOT NULL, password VARCHAR(40) NOT NULL, firstName VARCHAR(40) NOT NULL, lastName VARCHAR(40) NOT NULL, user_role role default \'users\', email VARCHAR(40) NOT NULL, UNIQUE(email, username) )');
+const request = client.query('CREATE TABLE requests(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), user_id UUID, title VARCHAR(40), currentStatus status default \'new\', details VARCHAR(255), UNIQUE(title), FOREIGN KEY(user_id) REFERENCES userlist(id) )');
+let count = 1;
+
+function endHandler() {
```
Comment (paradisekelechi) [inline server/models/database.js]: This is not how functions are defined using es6

File: server/models/database.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+dotenv.config();
+
+const connectionString = process.env.DATABASE_URL;
+
+const client = new pg.Client(connectionString);
+client.connect();
+const users = client.query('CREATE TABLE userlist(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), username VARCHAR(40) NOT NULL, password VARCHAR(40) NOT NULL, firstName VARCHAR(40) NOT NULL, lastName VARCHAR(40) NOT NULL, user_role role default \'users\', email VARCHAR(40) NOT NULL, UNIQUE(email, username) )');
+const request = client.query('CREATE TABLE requests(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), user_id UUID, title VARCHAR(40), currentStatus status default \'new\', details VARCHAR(255), UNIQUE(title), FOREIGN KEY(user_id) REFERENCES userlist(id) )');
+let count = 1;
```
Comment (paradisekelechi) [inline server/models/database.js]: I am not comfortable with this hardcoding here. What other ways could you implement this?

File: package.json (hunk 1 above)
Comment (cwizard2011) [reply package.json:34]: I am trying to figure it out, not sure

File: server/models/database.js (hunk 2 above)
Comment (cwizard2011) [reply server/models/database.js]: Thanks for the observation, I used it when I was finding it difficult to communicate with the database, I will rectify it now

File: server/models/database.js (hunk 5 above)
Comment (cwizard2011) [reply server/models/database.js]: To be sincere, I don't know any other way for now, but I will try and figure it out, I will appreciate if you can share a resource that will help

Comment (cwizard2011) [reply server/models/database.js]: I have been searching all through the night and couldn't find a better way. My peers that could have been of help have not reached this stage

File: package.json (hunk 1 above)
Comment (cwizard2011) [reply package.json:34]: I discovered that it has to do with semantic versioning scheme and its all about package update in general, I have updated my package.json

File: server/models/database.js (hunk 2 above)
Comment (cwizard2011) [reply server/models/database.js]: I have resolved this

File: server/models/database.js (hunk 4 above)
Comment (cwizard2011) [reply server/models/database.js]: I have rectified this

File: server/models/database.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+const connectionString = process.env.DATABASE_URL;
+
+const client = new pg.Client(connectionString);
+client.connect();
+client.query('CREATE TYPE role AS ENUM (\'users\', \'admin\')');
+client.query('CREATE TYPE status AS ENUM (\'new\', \'approved\', \'rejected\', \'pending\', \'resolved\')');
+const users = client.query('CREATE TABLE userlist(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), username VARCHAR(40) NOT NULL, password VARCHAR(40) NOT NULL, firstName VARCHAR(40) NOT NULL, lastName VARCHAR(40) NOT NULL, user_role role default \'users\', email VARCHAR(40) NOT NULL, UNIQUE(email, username) )');
+const request = client.query('CREATE TABLE requests(id UUID PRIMARY KEY DEFAULT gen_random_uuid(), user_id UUID, title VARCHAR(40), currentStatus status default \'new\', details VARCHAR(255), UNIQUE(title), FOREIGN KEY(user_id) REFERENCES userlist(id) )');
+let count = 2;
```
Comment (paradisekelechi) [inline server/models/database.js]: See the problem of hardcoding? 
You now have 4 queries whereas you are taking note of just two. The use of data structures would go a long way to help you

**Pull request merged**